checker/src/main/java/org/checkerframework/checker/index/Subsequence.java

```diff
@@ -0,0 +1,182 @@
+            array = standardizeAndViewpointAdapt(array, currentPath, context);
+        }
+
+        return new Subsequence(from, to, array);
+    }
+
+    /**
+     * Returns a Subsequence representing the {@link HasSubsequence} annotation on the declaration
+     * of {@code rec} or null if there is not such annotation.
+     *
+     * @param rec some tree
+     * @param factory AnnotatedTypeFactory
+     * @param currentPath the path at which to viewpoint adapt the subsequence
+     * @param context the context in which to viewpoint adapt the subsequence
+     * @return null or a new Subsequence from the declaration of {@code varTree}
+     */
+    public static Subsequence getSubsequenceFromReceiver(
+            Receiver rec,
+            AnnotatedTypeFactory factory,
+            TreePath currentPath,
+            FlowExpressionContext context) {
+        if (rec == null) {
+            return null;
+        }
+
+        Element element;
+        if (rec instanceof FieldAccess) {
+            element = ((FieldAccess) rec).getField();
+        } else if (rec instanceof LocalVariable) {
+            element = ((LocalVariable) rec).getElement();
+        } else {
+            return null;
+        }
+        return createSubsequence(
+                factory.getDeclAnnotation(element, HasSubsequence.class), currentPath, context);
+    }
+
+    /*
+     * Helper function to standardize and viewpoint adapt a String given a path and a context.
+     * Wraps FlowExpressionParseUtil#parse. If a parse exception is encountered, this returns
+     * its argument.
+     */
+    private static String standardizeAndViewpointAdapt(
+            String s, TreePath currentPath, FlowExpressionContext context) {
+        try {
+            s = FlowExpressionParseUtil.parse(s, context, currentPath, false).toString();
+        } catch (FlowExpressionParseException e) {
+        }
+        return s;
+    }
+
+    /**
+     * If the passed receiver is a FieldAccess or a LocalVariable, returns the context associated
+     * with it. Otherwise returns null.
+     *
+     * <p>Used to standardize and viewpoint adapt arguments to HasSubsequence annotations.
+     */
+    public static FlowExpressionContext getContextFromReceiver(Receiver rec, BaseContext checker) {
+        if (rec == null) {
+            return null;
+        }
+        FlowExpressionParseUtil.FlowExpressionContext context = null;
+        if (rec instanceof FlowExpressions.FieldAccess) {
+            FieldAccess fa = (FlowExpressions.FieldAccess) rec;
+            context =
+                    new FlowExpressionParseUtil.FlowExpressionContext(
+                            fa.getReceiver(), null, checker);
+
+        } else if (rec instanceof FlowExpressions.LocalVariable) {
+            LocalVariable lv = (FlowExpressions.LocalVariable) rec;
+            context = new FlowExpressionParseUtil.FlowExpressionContext(lv, null, checker);
```

Comment:
This isn't right for local variables.  The second parameter needs to be a list for formal parameters of the method in which the local variable is declared.  (You may just want to restrict `@HasSubsequence` to fields to make this simpler.)